README.md
# UOmapToolSet
Tools to modify map using python.


The main idea is to create a UOCoast.py similar to UOCoast.exe
But during the way I think I would like to create more little tools, and the idea is that anyone could use this as template to do their own tools or modifications.

I am not programmer, dont expect too much organization here or soft code. But I try.

Python is too slow dealing with maps. It can take more than one day to complete some tasks.


changelog.txt

[19/01/2023]
-Created base.py. This fille will read a map#.mul cell by cell and right now is used to developt and test base code.    
-NEw function->FindCoords(cellID)
-New function ->findCell(x , y) Find the cellId for given coords.
-Renamed Base.py to AllCells.py

-Removed all .py files.

-Added Interface.py
      This got all base function done until now and allow to test any change on it.

[21/01/2023]
-Added uocoast_.py
      -Able to check and draw 3x3 of any/all map cells.

-Added getpossiblesquares.py

[26/01/2023]
-ADDED  t-oceans.py Should be able to create transparent tiles. You can customize certain option. IT generates statics water on swc file. It is not tested yet.






Base.PY
#this is the earliest version of base tool. Right now it just read map and do prints to check it is able to read it properly and get cellid from coords or coords from cellid.


import struct
import math

mapHeightInBlocks = 512
blockSize = 64
rowSize = 8

def findCell(x , y):
    blockX = math.floor(x / rowSize)
    blockY = math.floor(y / rowSize)
    blockId = blockX*mapHeightInBlocks + blockY
    cellX= x - blockX*rowSize
    cellY=(y - blockY*rowSize)*8
    cellid= blockId*blockSize + cellX + cellY
    print(f"{cellid} Cellid the findcell")
    print(f"{blockId} BLOCKID")
    print(f"{blockX} BLOCKX")
    print(f"{blockY} BLOCKY")

def findCoords(cellId):
    blockId = math.floor(cellId / blockSize)

    blockX = math.floor(blockId / mapHeightInBlocks)
    blockY = blockId % mapHeightInBlocks

    cellX = cellId % rowSize
    cellY = math.floor( (cellId % blockSize) / rowSize )

    worldX = blockX * rowSize + cellX
    worldY = blockY * rowSize + cellY
    print("cellId: %s" % (cellId))
    print("world.x: %s" % (worldX))
    print("world.Y: %s" % (worldY))
    print("blockX: %s" % (blockX))
    print("blockY: %s" % (blockY))
    findCell(worldX , worldY)
        
cellId = -1
mapHeightInBlocks = 512
blockSize = 64
rowSize = 8
with open("map0.mul", "rb") as file:
    while True:
        cellId +=1
        data = file.read(3)
        if not data:
            break
        (tile_id, height) = struct.unpack("<Hb", data)
        findCoords(cellId)

    



Allcells.PY
#this is the earliest version of base tool. Right now it just read map and do prints to check it is able to read it properly and get cellid from coords or coords from cellid.
#thx Asylum as hard part of the code is from him.

import struct
import math

mapHeightInBlocks = 512
blockSize = 64
rowSize = 8

def findCell(x , y):
    blockX = math.floor(x / rowSize)
    blockY = math.floor(y / rowSize)
    blockId = blockX*mapHeightInBlocks + blockY
    cellX= x - blockX*rowSize
    cellY=(y - blockY*rowSize)*8
    cellid= blockId*blockSize + cellX + cellY
    print(f"{cellid} Cellid the findcell")
    print(f"{blockId} BLOCKID")
    print(f"{blockX} BLOCKX")
    print(f"{blockY} BLOCKY")

def findCoords(cellId):
    blockId = math.floor(cellId / blockSize)

    blockX = math.floor(blockId / mapHeightInBlocks)
    blockY = blockId % mapHeightInBlocks

    cellX = cellId % rowSize
    cellY = math.floor( (cellId % blockSize) / rowSize )

    worldX = blockX * rowSize + cellX
    worldY = blockY * rowSize + cellY
    print("cellId: %s" % (cellId))
    print("world.x: %s" % (worldX))
    print("world.Y: %s" % (worldY))
    print("blockX: %s" % (blockX))
    print("blockY: %s" % (blockY))
    findCell(worldX , worldY)
        
cellId = -1
mapHeightInBlocks = 512
blockSize = 64
rowSize = 8
with open("map0.mul", "rb") as file:
    while True:
        cellId +=1
        data = file.read(3)
        if not data:
            break
        (tile_id, height) = struct.unpack("<Hb", data)
        findCoords(cellId)

    



T-Oceans.PY

import os
import struct
import math
import tkinter as tk
from tkinter import messagebox
from tkinter import ttk
import random

mapHeightInBlocks = 512
blockSize = 64
rowSize = 8
mapwidth = 7168 #set ur map dimensions here.
mapheight = 4096
floorCave = [581, 582, 583, 584, 585] #Change this for another terrain for the bottom.
staticWater = [6040, 6041, 6042, 6043, 6044] #change this if custom water tiles.

def f_iswater(tileid):
    if (tileid >= 168 and tileid <= 171) or \
       (tileid >= 310 and tileid <= 311) or \
       (tileid >= 22455 and tileid <= 22460) or \
       (tileid >= 29902 and tileid <= 29917) or \
       (tileid >= 29952 and tileid <= 29975) or \
       (tileid >= 29981 and tileid <= 30013):
        return 1
    else:
        return -1

    
def findCell(x , y):
    blockX = math.floor(x / rowSize)
    blockY = math.floor(y / rowSize)
    blockId = blockX*mapHeightInBlocks + blockY
    cellX= x - blockX*rowSize
    cellY=(y - blockY*rowSize)*8
    cellid= blockId*blockSize + cellX + cellY
    return (cellid)

def findCellxy(x , y):
    blockX = math.floor(x / rowSize)
    blockY = math.floor(y / rowSize)
    blockId = blockX*mapHeightInBlocks + blockY
    cellX= x - blockX*rowSize
    cellY=(y - blockY*rowSize)
    return (cellX, cellY)


def findCoords(cellId):
    blockId = math.floor(cellId / blockSize)

    blockX = math.floor(blockId / mapHeightInBlocks)
    blockY = blockId % mapHeightInBlocks

    cellX = cellId % rowSize
    cellY = math.floor( (cellId % blockSize) / rowSize )

    worldX = blockX * rowSize + cellX
    worldY = blockY * rowSize + cellY
    ccellid=findCell(worldX , worldY)
    return (worldX , worldY)
    

def getTerrain(cellId):
    with open("map0.mul", "rb") as filemap0:
        headersfix = math.floor(cellId/64)*4
        biites = cellId
        filemap0.seek(4+headersfix+biites*3)   
        data = filemap0.read(3)
        (tile_id, height) = struct.unpack("<Hb", data)
        return tile_id
        
def readAll():
    cellId = -1 #set here the starting cell.
    with open("map0.mul", "r+b") as filemap0:
        while cellId < mapwidth*mapheight:
            cellId +=1    
            headersfix = math.floor(cellId/64)*4
            biites = cellId
            filemap0.seek(4+headersfix+biites*3)   
            data = filemap0.read(3)
            (tile_id, height) = struct.unpack("<Hb", data)
            if f_iswater(tile_id)==1:
                tile_id=random.choice(floorCave)
                (X2 , Y2)=findCoords(cellId)
                (X , Y)=findCellxy(X2 , Y2)
                if Y <=3:
                    fix = Y
                elif Y >=4:
                    fix = 8-Y
                hheight=height - 20 - fix
                new_data = struct.pack("<Hb", tile_id, hheight)
                # Volver al lugar donde se encontraban los datos antiguos
                filemap0.seek(4+headersfix+biites*3)
                # Escribir los nuevos datos en el archivo
                filemap0.write(new_data)
                print(cellId)
                ID=random.choice(staticWater)
                with open("statics.txt", "a") as file:
                    file.write(F"SECTION WORLDITEM {cellId}" + "\n")
                    file.write("{" + "\n")
                    file.write(f"SERIAL {cellId}" + "\n")
                    file.write(f"ID {ID}" + "\n")
                    file.write(f"X {X2}" + "\n")
                    file.write(f"Y {Y2}" + "\n")
                    file.write(f"Z {height}" + "\n")
                    file.write("TYPE 255" + "\n")
                    file.write("COLOR 0" + "\n")
                    file.write("}" + "\n")

def readRect(x1 , y1, x2 , y2):
    #with open("map0.mul", "rb") as file:
        #for
    print("not yet")
        
ventana = tk.Tk()
ventana.title("MapTester")
ventana.geometry('400x125')
entry1 = ttk.Entry()
entry1.insert(0, "CellId")
entry1.place(x=10, y=0)
entry2 = ttk.Entry()
entry2.insert(0, "x1")
entry2.place(x=10, y=25)
entry3 = ttk.Entry()
entry3.insert(0, "y1")
entry3.place(x=10, y=50)                
entry4 = ttk.Entry()
entry4.insert(0, "x2")
entry4.place(x=10, y=75)
entry5 = ttk.Entry()
entry5.insert(0, "y2")
entry5.place(x=10, y=100)

#boton1 = tk.Button(ventana, text="SingleCell", command=lambda : drawBoards(int(entry1.get())))
#boton1.pack()
#boton2 = tk.Button(ventana, text="ReadCoords x1 y1", command=lambda : readSingle(int(entry1.get())))
#boton2.pack()
boton3 = tk.Button(ventana, text="READ all Cells", command=readAll)
boton3.pack()

ventana.mainloop()





Interface.PY

import os
import struct
import math
import tkinter as tk
from tkinter import messagebox
from tkinter import ttk

mapHeightInBlocks = 512
blockSize = 64
rowSize = 8

def findCell(x , y):
    blockX = math.floor(x / rowSize)
    blockY = math.floor(y / rowSize)
    blockId = blockX*mapHeightInBlocks + blockY
    cellX= x - blockX*rowSize
    cellY=(y - blockY*rowSize)*8
    cellid= blockId*blockSize + cellX + cellY
    print(f"{cellid} Cellid the findcell")
    print(f"{blockId} BLOCKID")
    print(f"{blockX} BLOCKX")
    print(f"{blockY} BLOCKY")

def findCoords(cellId):
    blockId = math.floor(cellId / blockSize)

    blockX = math.floor(blockId / mapHeightInBlocks)
    blockY = blockId % mapHeightInBlocks

    cellX = cellId % rowSize
    cellY = math.floor( (cellId % blockSize) / rowSize )

    worldX = blockX * rowSize + cellX
    worldY = blockY * rowSize + cellY
    print("cellId: %s" % (cellId))
    print("world.x: %s" % (worldX))
    print("world.Y: %s" % (worldY))
    print("blockX: %s" % (blockX))
    print("blockY: %s" % (blockY))
    findCell(worldX , worldY)
        
def readSingle(cellId):
    with open("map0.mul", "rb") as file:
        biites = cellId
        file.seek(biites)    # move the file pointer forward 6 bytes (i.e. to the 'w')
        data = file.read(3)
        (tile_id, height) = struct.unpack("<Hb", data)
        findCoords(cellId)
        
def readAll():
    cellId = -1
    with open("map0.mul", "rb") as file:
        while True:
            cellId +=1
            data = file.read(3)
            if not data:
                break
            (tile_id, height) = struct.unpack("<Hb", data)
            findCoords(cellId)

def readRect(x1 , y1, x2 , y2):
    #with open("map0.mul", "rb") as file:
        #for
    print("hello world")

def salir():
    salir = messagebox.askyesno("Salir", "¿Estás seguro de que deseas salir?")
    if salir:
        ventana.destroy()
        
ventana = tk.Tk()
ventana.title("MapTester")
ventana.geometry('400x125')
entry1 = ttk.Entry()
entry1.insert(0, "CellId")
entry1.place(x=10, y=0)
entry2 = ttk.Entry()
entry2.insert(0, "x1")
entry2.place(x=10, y=25)
entry3 = ttk.Entry()
entry3.insert(0, "y1")
entry3.place(x=10, y=50)
entry4 = ttk.Entry()
entry4.insert(0, "x2")
entry4.place(x=10, y=75)
entry5 = ttk.Entry()
entry5.insert(0, "y2")
entry5.place(x=10, y=100)

boton1 = tk.Button(ventana, text="SingleCell", command=lambda : readSingle(int(entry1.get())))
boton1.pack()
boton2 = tk.Button(ventana, text="ReadCoords x1 y1", command=lambda : readSingle(int(entry1.get())))
boton2.pack()
boton3 = tk.Button(ventana, text="READ all Cells", command=readAll)
boton3.pack()
boton4 = tk.Button(ventana, text="Read rectangle", command=lambda : readRect(0 ,0 ,12 ,12))
boton4.pack()
boton9 = tk.Button(ventana, text="QUIT", command=salir)
boton9.pack()

ventana.mainloop()   



SingleCell.PY
import struct
import math

mapHeightInBlocks = 512
blockSize = 64
rowSize = 8

def findCell(x , y):
    blockX = math.floor(x / rowSize)
    blockY = math.floor(y / rowSize)
    blockId = blockX*mapHeightInBlocks + blockY
    cellX= x - blockX*rowSize
    cellY=(y - blockY*rowSize)*8
    cellid= blockId*blockSize + cellX + cellY
    print(f"{cellid} Cellid the findcell")
    print(f"{blockId} BLOCKID")
    print(f"{blockX} BLOCKX")
    print(f"{blockY} BLOCKY")

def findCoords(cellId):
    blockId = math.floor(cellId / blockSize)

    blockX = math.floor(blockId / mapHeightInBlocks)
    blockY = blockId % mapHeightInBlocks

    cellX = cellId % rowSize
    cellY = math.floor( (cellId % blockSize) / rowSize )

    worldX = blockX * rowSize + cellX
    worldY = blockY * rowSize + cellY
    print("cellId: %s" % (cellId))
    print("world.x: %s" % (worldX))
    print("world.Y: %s" % (worldY))
    print("blockX: %s" % (blockX))
    print("blockY: %s" % (blockY))
    findCell(worldX , worldY)
        
#def readsingle(cellId)
cellId=128
with open("map0.mul", "rb") as file:
        biites = cellId
        file.seek(biites)    # move the file pointer forward 6 bytes (i.e. to the 'w')
        data = file.read(3)
        (tile_id, height) = struct.unpack("<Hb", data)
        findCoords(cellId)

    



UOCoast_.PY

import os
import struct
import math
import tkinter as tk
from tkinter import messagebox
from tkinter import ttk

mapHeightInBlocks = 512
blockSize = 64
rowSize = 8
board_Cells = [[0 for i in range(3)] for j in range(3)]
board_height = [[0 for i in range(3)] for j in range(3)]
board_terrain = [[0 for i in range(3)] for j in range(3)]
water_tiles = ["hi", "hello", "goodbye"]
mapwidth = 7168 #set ur map dimensions here.
mapheight = 4096

def f_iswater(tileid):
    #print(tileid)
    #print(hex(tileid))
    if (tileid >= 0x0a8 and tileid <= 0x0ab) or \
       (tileid >= 0x136 and tileid <= 0x137) or \
       (tileid >= 0x5797 and tileid <= 0x579c) or \
       (tileid >= 0x746e and tileid <= 0x7485) or \
       (tileid >= 0x7490 and tileid <= 0x74ab) or \
       (tileid >= 0x74b5 and tileid <= 0x75d5):
        return 1
    else:
        return -1

    
def findCell(x , y):
    blockX = math.floor(x / rowSize)
    blockY = math.floor(y / rowSize)
    blockId = blockX*mapHeightInBlocks + blockY
    cellX= x - blockX*rowSize
    cellY=(y - blockY*rowSize)*8
    cellid= blockId*blockSize + cellX + cellY
    return (cellid)

def findCoords(cellId):
    blockId = math.floor(cellId / blockSize)

    blockX = math.floor(blockId / mapHeightInBlocks)
    blockY = blockId % mapHeightInBlocks

    cellX = cellId % rowSize
    cellY = math.floor( (cellId % blockSize) / rowSize )

    worldX = blockX * rowSize + cellX
    worldY = blockY * rowSize + cellY
    ccellid=findCell(worldX , worldY)
    return (worldX , worldY)
    

def getTerrain(cellId):
    with open("map0.mul", "rb") as filemap0:
        headersfix = math.floor(cellId/64)*4
        biites = cellId
        filemap0.seek(4+headersfix+biites*3)   
        data = filemap0.read(3)
        (tile_id, height) = struct.unpack("<Hb", data)
        return tile_id

def getheight(cellId):
    with open("map0.mul", "rb") as file:
        biites = cellId
        file.seek(biites)
        data = file.read(3)
        (tile_id, height) = struct.unpack("<Hb", data)
        return height

def checksquare():
    global board_terrain
    squareId = ''.join(str(i) for i in board_terrain)
    squareId = "".join([i if i in ["G", "W"] else "" for i in squareId])
    print(squareId)
    if "X" in squareId or squareId.find("W") != -1
        return
    if squareId[4] == "G": #we are working on Ground 
        if squareId == "WWWGGGGGG":
            print("north'.")
        elif squareId == "GGGGGGWWW":
            print("south")
        elif squareId == "GGWGGWGGW":
            print("east")
        elif squareId == "WGGWGGWGG":
            print("west")
        elif squareId == "WGGWGGWGG":
            print("hmmm need think about this")

    elif squareId[4] == "W": #we are working on water. 
        if squareId == "WWWGWGGGG":
            print("north'.")
        elif squareId == "GGGGWGWWW":
            print("south")
        elif squareId == "GGWGWWGGW":
            print("east")
        elif squareId == "WGGWWGWGG":
            print("west")
        elif squareId == "WGGWWGWGG":
            print("hmmm need think about this")


def drawBoards(cellId):
    (mx, my) = findCoords(cellId)
    print (mx, my)
    center = 1
    global board_terrain
    global board_Cells
    global board_height
    for x in range(3):
        for y in range(3):
            xx = x-center+my
            yy = y-center+mx
            #print(f"Printingxxyy ({xx} {yy})")
            #print(f"Printingxy ({x} {y})")
            if (xx <= -1 or yy <= -1) or (xx >= mapwidth or yy >= mapheight):
                board_terrain[x][y] = "X" #outside map
                continue
            ccellid=findCell(xx, yy)
            tileid = getTerrain(ccellid)
                
            if f_iswater(tileid)==1:
                board_terrain[x][y] = "W"
            else:
                board_terrain[x][y] = "G" 
            #board_Cells[x][y] = ccellid
            #board_height[x][y] = getheight(ccellid)
    print(f"Printing 3x3 of {cellId} pos ({mx} {my})")
    for row in board_terrain:
        for cell in row:
            print(cell, end='')
        print() # para saltar de linea
    print (board_terrain)

def getTerrainXY(x , y):
    cellId = findCell(x , y)
    terrain = readSingle(cellId)
    return terrain

        
def readAll():
    cells = -1
    with open("map0.mul", "rb") as file:
        while cells < mapwidth*mapheight:
            cells +=1
            drawBoards(cells)

def readRect(x1 , y1, x2 , y2):
    #with open("map0.mul", "rb") as file:
        #for
    print("hello world")

def salir():
    salir = messagebox.askyesno("Salir", "¿Estás seguro de que deseas salir?")
    if salir:
        ventana.destroy()

########################################
        
def check_adjacent(board, row, col):
    rows = len(board)
    cols = len(board[0])
    adjacent = []
    # Check top
    if row > 0:
        adjacent.append(board[row-1][col])
    # Check bottom
    if row < rows-1:
        adjacent.append(board[row+1][col])
    # Check left
    if col > 0:
        adjacent.append(board[row][col-1])
    # Check right
    if col < cols-1:
        adjacent.append(board[row][col+1])
    # Check top-left
    if row > 0 and col > 0:
        adjacent.append(board[row-1][col-1])
    # Check top-right
    if row > 0 and col < cols-1:
        adjacent.append(board[row-1][col+1])
    # Check bottom-left
    if row < rows-1 and col > 0:
        adjacent.append(board[row+1][col-1])
    # Check bottom-right
    if row < rows-1 and col < cols-1:
        adjacent.append(board[row+1][col+1])
    return adjacent



def convert_to_rgb(color):
    
    red = (color >> 11) & 0x1F
    green = (color >> 5) & 0x3F
    blue = color & 0x1F
    
    red = (red << 3) | (red >> 2)
    green = (green << 2) | (green >> 4)
    blue = (blue << 3) | (blue >> 2)
    
    return (red, green, blue)

    
#board = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
#board[0][0] = (x, y) = findCoords(cellId)
#print(check_adjacent(board, 1, 1)) # prints [2, 4, 6, 8]

with open("RadarCol.mul", "rb") as radarcol:
    radar_data = radarcol.read()
colors = struct.unpack("<" + "H" * int(len(radar_data) / 2), radar_data)        
ventana = tk.Tk()
ventana.title("MapTester")
ventana.geometry('400x125')
entry1 = ttk.Entry()
entry1.insert(0, "CellId")
entry1.place(x=10, y=0)
entry2 = ttk.Entry()
entry2.insert(0, "x1")
entry2.place(x=10, y=25)
entry3 = ttk.Entry()
entry3.insert(0, "y1")
entry3.place(x=10, y=50)                
entry4 = ttk.Entry()
entry4.insert(0, "x2")
entry4.place(x=10, y=75)
entry5 = ttk.Entry()
entry5.insert(0, "y2")
entry5.place(x=10, y=100)

boton1 = tk.Button(ventana, text="SingleCell", command=lambda : drawBoards(int(entry1.get())))
boton1.pack()
boton2 = tk.Button(ventana, text="ReadCoords x1 y1", command=lambda : readSingle(int(entry1.get())))
boton2.pack()
boton3 = tk.Button(ventana, text="READ all Cells", command=readAll)
boton3.pack()
boton4 = tk.Button(ventana, text="Read rectangle", command=lambda : readRect(0 ,0 ,12 ,12))
boton4.pack()
boton9 = tk.Button(ventana, text="QUIT", command=salir)
boton9.pack()

ventana.mainloop()



